### PR TITLE
Swap from exact substring matches to each individual word matching

### LINKED ISSUE #143: Implement fuzzier search for model names in docs
### Describe the feature
Searching for model names should exclude underscores and other word break characters, and be able to find tokens in any order. 

### Who will this benefit?
We (like probably everyone else) use snake_case in our model names. When looking for a model in dbt Docs, I try to search by a couple of keywords (e.g. "class teachers") if I know a model exists but don't remember its exact name. This doesn't find a model called `teachers__class_counts`, because the search appears to require a perfect substring match, such as `_class_c`. 
image

Even if I remember the model name, having to remember where the underscores go is a pain - is it `user_meta_data` or `user_metadata`?

### Describe alternatives you've considered
* Having a better memory
* Searching for a single word using the dbt search bar, then searching for a second word using Ctrl-F, e.g. 
image

### Additional context
If separator characters are included in the search then they shouldn't be stripped. Perfect substring matches should _probably_ still win, but I haven't put a lot of thought into those edge cases. 

### Are you interested in contributing this feature?
My JS is pretty much limited to hello world, but if you point me in the right direction I'm happy to lose an hour or two trying before giving up.

### PR DESCRIPTION
**Pull request by joellabes**
resolves #143 

### Description
This is a first attempt at considering each word entered in a search individually, as opposed to requiring a perfect match (`model my` or `my model` should now find `my_model`, where earlier only an exact search, including underscores, would have worked). `my_model` still works of course.

I couldn't get the Docs site running locally (from looking at the stack traces, seemed to be a bunch of issues with webpack, node versions, python versions...) but ~I have tested that each individual piece of work behaves correctly outside of the wider Docs context~ I tried it in the Netlify CI preview and it works correctly.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have generated docs locally, and this change appears to resolve the stated issue
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: src/app/services/project_service.js [hunk 1]
```diff
@@ -271,23 +271,24 @@ angular
             'tags': 'array',
             'arguments': 'array',
         };
-        var search = new RegExp(val, "i")
```
Comment (joellabes) [inline src/app/services/project_service.js:274]: This was never used

**Review by jtcohen6 — COMMENTED** [review]
This is really cool @joellabes !!

I'll wait until @drewbanin can give the JS a once-over, likely next week

**Comment by drewbanin** [comment]
ha - oops - thanks for your patience @joellabes - giving this a look now :D

**Review by drewbanin — CHANGES_REQUESTED** [review]
I love what you're going for here, but I think it's about time that we giving searching some more love! What do you think about leveraging something like [fusejs](https://fusejs.io/demo.html) here and deleting a lot of this searching logic? I kicked around fuse.js, and it looks to me like it will match the query from the original issue (#143) out of the box!

Joel, if this is something you're interested in pursuing, I'd be really happy to offer some guidance! If not, I imagine it's something that we could get to in the reasonably near future. Let me know either way!

**Comment by joellabes** [comment]
I'd be so excited for search to get some love - I had an urge to rework the weighting process last week as well. Happy to have a blast with Fuse. I had dramas getting dbt running locally (Windows 🙄) but will give it another whirl - if it doesn't magically work I might have to get a hand on the initial setup!

**Comment by joellabes** [comment]
I've started looking into this on #162 - still a long way from it being ready, but I'm going to close this one for now.

**Comment by joellabes** [comment]
OK so 162 wound up being non-performant _and_ not very good at fuzzy search. Not a good combo. 

Since this has been sitting around waiting for a year and a bit, and works just fine, I'm going to reopen it and hopefully we can get it out into a future docs release. If someone who knows more JS wants to have a swing at fuse or some other proper search lib, I'm into it. 

But in general, dbt docs searches are only a couple of words and it's not unreasonable to just require accurate typing 😅

**Comment by joellabes** [comment]
@clablot check

**Review by stu-k — APPROVED** [review]
Just a couple comments about using lodash functions, nothing else stands out 👍

**Comment by emmyoop** [comment]
Tested this out and it seems like a big quality of life improvement! 🙌   I'll take it for another spin once you implement @stu-k's suggestions.  

Your commit from October 13, 2020 is causing the cla bot to fail because it's associated with your `joel.labes@educationperfect.com` email.  You should be able to squash down your commits and get the cla bot to pass.

File: src/app/components/search/search.js [hunk 2]
```diff
@@ -105,16 +113,27 @@ angular
...
-                return scope.limit_columns[id] !== undefined? scope.limit_columns[id] : 3;
+                return scope.limit_columns[id] !== undefined ? scope.limit_columns[id] : 3;
+            }
+
+            //from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
+            function escapeRegExp(string) {
+                return string.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
+            }
+
+            function splitQuery(query){
```
[the remark this answers is not in the document]
Comment (joellabes) [reply src/app/components/search/search.js:135]: Yep that's a better name! Have kept the function for lowercasing purposes but have swapped out all the icky split stuff for `_.words`

**Comment by joellabes** [comment]
@cla-bot check

**Comment by joellabes** [comment]
@emmyoop I've applied @stu-k's suggestions now 🛠️

I couldn't get the squash to work, so just re-added the educationperfect domain to my GH account and the CLA bot is happy!

**Review by emmyoop — APPROVED** [review]
Great work!  Thanks for staying on it and pushing it through.  The search improvements are so so satisfying.

**Pull request merged**